Snowflake.Data/Core/Session/SessionPool.cs

```diff
@@ -97,8 +101,93 @@ internal SFSession GetSession(string connStr)
                     }
                 }
             }
-            return null;
+            return NewSession(connStr, password);
         }
+        
+        internal Task<SFSession> GetSessionAsync(string connStr, SecureString password, CancellationToken cancellationToken)
+        {
+            s_logger.Debug("SessionPool::GetSession");
+            if (!_pooling)
+                return NewSessionAsync(connStr, password, cancellationToken);
+            lock (s_sessionPoolLock)
+            {
+                for (int i = 0; i < _sessionPool.Count; i++)
+                {
+                    if (_sessionPool[i].connStr.Equals(connStr))
+                    {
+                        SFSession session = _sessionPool[i];
+                        _sessionPool.RemoveAt(i);
+                        long timeNow = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
+                        if (session.IsExpired(_timeout, timeNow))
+                        {
+                            session.close();
+                            i--;
+                        }
+                        else
+                        {
+                            s_logger.Debug($"reuse pooled session with sid {session.sessionId}");
+                            return Task.FromResult(session);
+                        }
+                    }
+                }
+            }
+            return NewSessionAsync(connStr, password, cancellationToken);
+        }
+
+        internal SFSession NewSession(String connectionString, SecureString password)
+        {
+            try
+            {
+                var session = new SFSession(connectionString, password);
+                session.Open();
+                return session;
+            }
+            catch (Exception e)
+            {
+                // Otherwise when Dispose() is called, the close request would timeout.
+                if (e is SnowflakeDbException)
+                    throw;
+                throw new SnowflakeDbException(
+                    e,
+                    SnowflakeDbException.CONNECTION_FAILURE_SSTATE,
+                    SFError.INTERNAL_ERROR,
+                    "Unable to connect. " + e.Message);
+            }
+        }
+
+        internal Task<SFSession> NewSessionAsync(String connectionString, SecureString password, CancellationToken cancellationToken)
+        {
+            try
+            {
+                var session = new SFSession(connectionString, password);
+                return session
+                    .OpenAsync(cancellationToken)
+                    .ContinueWith(previousTask =>
+                    {
+                        if (previousTask.IsFaulted && previousTask.Exception != null)
```

Comment:
isn't this logic the same as in line 181?
If we would remove this ContinueWith code wouldn't the result be the same?
Maybe it is fine, but I'm just asking.

Reply:
ContinueWith is crucial as we need to return session out of async call.
OpenAsync returns future task with just a state. 
So catch handles any exception from constructor of a session.
ContinueWith needs to handle exceptions from OpenAsync as well as provide final opened session as a result.